### DISCUSSION TITLE
Is model_dump_json deterministic?

### DISCUSSION THREAD

**Discussion by sustefil** [pydantic/pydantic#10343] [Question]
Hello!

I'd like to ask whether it is safe to assume that the `model_dump_json` function is stable/deterministic so it always produces the same serialized JSON for the same data in a model? I'd like to use a hash function to determine the model uniqueness in my app. I couldn't find it in the docs.

Thanks!

**Comment by titusz** [comment]
I don´t know if `model_dump_json` is deterministic, but you could use [JCS canonicalization](https://pypi.org/project/jcs/) to achieve that.

```python
import jcs

deterministic = jcs.canonicalize(dict(pydantic_object))
```

**Comment by erklem** [comment]
It looks like the field serialization order is consistent with the order they're declared in the model. https://github.com/pydantic/pydantic/issues/7424#issuecomment-2480525714

There are other factors (e.g. whitespace) that could still affect determinism.

**Comment by vinicq** [comment]
Not all the way, and the gaps bite when you use the JSON bytes as a hash input or cache key.

What is stable across runs. Field order follows the declaration order on the model, and the separators don't drift between calls. erklem already linked the underlying behavior on field order.

The gaps I've hit in practice:

- Numeric types. Two semantically equal numbers can serialize differently if one entered the model as `Decimal` and another as `float`. Normalize the types before validation or use `Decimal` end to end.
- `datetime` with tzinfo. The output includes the offset, so a naive datetime and an explicit-UTC datetime that represent the same instant do not produce the same bytes.
- Optional fields with defaults. Not-set vs set-to-the-default produce different output unless `exclude_defaults` or `exclude_unset` is applied consistently.
- Free-form dict fields (`dict[str, Any]`). Field order on the model doesn't help here. The inner dict serializes in insertion order from whatever produced it.

For hashing or signature paths I treat `model_dump_json` as not canonical by default and pipe through a canonicalizer. JCS is one option. If you'd rather not pull a dependency, `model_dump(mode="json")` fed into `json.dumps(..., sort_keys=True, separators=(",", ":"))` gets you a deterministic byte string with no extras.

If byte stability matters for HMAC or content addressing, write the canonical form once and snapshot test the bytes. Catches the day someone adds an Optional field with a default and silently changes the hash for every existing record.

**Comment by ltianyi992** [comment]
Yes, `model_dump_json()` is deterministic for the same model and data.

**Why it is stable:**

1. **Field order follows definition order** — Pydantic v2 serializes fields in the order they are defined on the model, using Rust-backed serialization (`pydantic-core`). This order is fixed at class-creation time and does not change between calls.

2. **Python dicts are ordered since 3.7** — nested dict-typed fields serialize in insertion order consistently.

3. **The Rust serializer is deterministic** — the same input produces the same JSON bytes on every call within a process (and across processes/platforms for the same Python + pydantic-core version).

**Practical caveat — floats:**
Float serialization can differ very slightly between Python versions or platforms for certain edge-case values (e.g., very large/small numbers near float precision limits). For integer, string, bool, None, and datetime fields, output is identical.

**For hashing model uniqueness**, this is a common and safe pattern:

```python
import hashlib
from pydantic import BaseModel

class MyModel(BaseModel):
    name: str
    value: int

def model_hash(m: MyModel) -> str:
    return hashlib.sha256(m.model_dump_json().encode()).hexdigest()

a = MyModel(name="x", value=1)
b = MyModel(name="x", value=1)
assert model_hash(a) == model_hash(b)  # True, deterministic
```

**One thing to watch:** if your model uses `computed_field` or `default_factory` that produces non-deterministic output (e.g., `uuid4()`, `datetime.now()`), those will obviously vary. The serializer itself is stable; non-determinism only enters from the data.
